### DISCUSSION TITLE
Add support for azure cloud in trivy

### DISCUSSION THREAD

**Discussion by AkhtarAmir** [aquasecurity/trivy#4519] [Development]
Today we support aws cloud scanning in trivy. The intention is to support scanning for Azure cloud as well.

**Comment by simar7** [comment]
Thanks for starting the discussion @AkhtarAmir. Can you please describe what the user experience would look like? I assume the following:

```shell
$ trivy azure
```

Would scan everything in your azure cloud? 

Similar options would work for azure as they do today, for example: `--services=<service-name>` would target just that particular service?

Feel free to add more details here.

**Reply by AkhtarAmir** [reply]
Yeah so it will be very similar to what we do with aws today. It will take required creds like:
```
subscription_id: "",
key_value: "",
directory_id: ""
application_id: ""
```
 from user as we do in CSPM currently and scan azure cloud account.
Now, how will we manage this input is debatable. The two most obvious ways are cmd arguments and from config file. What do you think @itaysk @simar7

**Reply by simar7** [reply]
For the credentials as input, we should closely follow how Trivy AWS does it today, which in turn follows the `aws cli` model. Which is https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html#configure-precedence

Or in short (relevant to Trivy):

1. Command line options
2. Environment variables
3. Credentials on disk (`.aws/credentials`) or similar.

For Azure, the list loosely follows the above and is described here https://learn.microsoft.com/en-us/cli/azure/azure-cli-configuration

**Reply by simar7** [reply]
AFAIK with aws, this precedence order is already set and usable via the AWS SDK. So users of the AWS SDK don't have to implement this. I assume it should be the case with Azure SDK as well but this behaviour should be verified.

**Reply by AkhtarAmir** [reply]
So Azure SDK does not save creds locally, instead it creates access_token and refresh_token so we can't replicate AWS behaviour. I was thinking taking config.json as azure flag and parse creds from it

**Reply by simar7** [reply]
As discussed offline, @AkhtarAmir will investigate further into how Azure SDK handles auth.

**Reply by knqyf263** [reply]
I'm unsure if it helps, but we recently did something around Azure auth.
https://github.com/aquasecurity/trivy/pull/4489

**Reply by AkhtarAmir** [reply]
It is helpful. Thanks
